src/main/java/com/microsoft/graph/serializer/CollectionResponseSerializer.java

```diff
@@ -0,0 +1,85 @@
+import javax.annotation.Nullable;
+
+import com.google.gson.JsonArray;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParseException;
+import com.microsoft.graph.http.BaseCollectionResponse;
+import com.microsoft.graph.logger.ILogger;
+
+/** Specialized serializer to handle collection responses */
+public class CollectionResponseSerializer {
+    private static DefaultSerializer serializer;
+    /**
+     * Not available for instantiation
+     */
+    private CollectionResponseSerializer() {}
+    /**
+     * Deserializes the JsonElement
+     *
+     * @param json the source CollectionResponse's Json
+     * @param typeOfT The type of the CollectionResponse to deserialize to
+     * @param logger the logger
+     * @param <T1> the entity type for the collection
+     * @throws JsonParseException the parse exception
+     * @return    the deserialized CollectionResponse
+     */
+    @SuppressWarnings("unchecked")
+    @Nullable
+    public static <T1> BaseCollectionResponse<T1> deserialize(@Nonnull final JsonElement json, @Nonnull final Type typeOfT, @Nonnull final ILogger logger) throws JsonParseException {
+        if (json == null || !json.isJsonObject()| !typeOfT.getClass().equals(Class.class)) {
+            return null;
+        }
+        serializer = new DefaultSerializer(logger);
+        final JsonObject jsonAsObject = json.getAsJsonObject();
+        final JsonArray sourceArray = jsonAsObject.get("value").getAsJsonArray();
+        final ArrayList<T1> list = new ArrayList<T1>(sourceArray.size());
+        /** eg: com.microsoft.graph.requests.AttachmentCollectionResponse */
+        /** eg: com.microsoft.graph.requests.DriveItemDeltaCollectionResponse */
+        /** eg: com.microsoft.graph.requests.DriveItemGetActivitiesByIntervalCollectionResponse */
+        final Class<?> responseClass = ((Class<?>)typeOfT);
+        /** eg: com.microsoft.graph.http.BaseCollectionResponse<com.microsoft.graph.models.DriveItem> */
+        final String genericResponseName = responseClass.getGenericSuperclass().toString();
+        final int indexOfGenericMarker = genericResponseName.indexOf('<');
+        /** eg: com.microsoft.graph.models.Attachment */
+        final String baseEntityClassCanonicalName = genericResponseName
+                                                        .substring(indexOfGenericMarker + 1, genericResponseName.length() - 1);
+        Class<?> baseEntityClass;
+        try {
+            baseEntityClass = Class.forName(baseEntityClassCanonicalName);
+        } catch (ClassNotFoundException ex) {
+            //it is possible we can't find the parent base class depending on the response class name, see examples
+            baseEntityClass = null;
+            logger.logDebug("could not find class" + baseEntityClassCanonicalName);
+        }
+        try {
+            for(JsonElement sourceElement : sourceArray) {
+                if(sourceElement.isJsonObject()) {
```

Comment:
The sourceElement is skipped if this is a jsonArray or JsonProperty. I don't know of any current response collections that return a mix of json token types, but we may want to log a warning if `!sourceElement.isJsonObject()` in case a future workload has a mixed return type. Non-collection responses do return a mix of jsonObject, jsonProperty and jsonArray.